iscsi/get.go

```diff
@@ -0,0 +1,283 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package iscsi
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+)
+
+// GetStats is the main iscsi status information func
+// building the path and prepare info for enable iscsi
+func GetStats(iqnPath string) (*Stats, error) {
+	var istats Stats
+
+	istats.Name = filepath.Base(iqnPath)
+	istats.RootPath = filepath.Dir(iqnPath)
+
+	matches, err := filepath.Glob(filepath.Join(iqnPath, "tpgt*"))
+	if err != nil {
+		return nil, fmt.Errorf("iscsi: GetStats: get TPGT path error %v", err)
+	}
+	istats.Tpgt = make([]TPGT, len(matches))
+
+	for i, tpgtPath := range matches {
+		istats.Tpgt[i].Name = filepath.Base(tpgtPath)
+		istats.Tpgt[i].TpgtPath = tpgtPath
+		istats.Tpgt[i].IsEnable, _ = isPathEnable(tpgtPath)
+		if istats.Tpgt[i].IsEnable {
+			matchesLunsPath, _ := filepath.Glob(filepath.Join(tpgtPath, "lun/lun*"))
+
+			for _, lunPath := range matchesLunsPath {
+				lun, err := getLunLinkTarget(lunPath)
+				if err == nil {
+					istats.Tpgt[i].Luns = append(istats.Tpgt[i].Luns, lun)
+				}
+			}
+		}
+	}
+	return &istats, nil
+}
+
+// isPathEnable is a utility function
+// check if the file "enable" contain enable message
+func isPathEnable(path string) (bool, error) {
+	var isEnable bool
+	isEnable = false
+	enableReadout, err := ioutil.ReadFile(filepath.Join(path, "enable"))
+	if err != nil {
+		return false, fmt.Errorf("iscsi: isPathEnable error %v", err)
+	}
+	enableReadoutNum, err := strconv.Atoi(strings.TrimSpace(string(enableReadout)))
+	if err != nil {
+		return false, fmt.Errorf("Atoi fail %v", err)
+	}
+	if enableReadoutNum > 0 {
+		isEnable = true
+	}
+	return isEnable, nil
+}
+
+func getLunLinkTarget(lunPath string) (lunObject LUN, err error) {
+	lunObject.Name = filepath.Base(lunPath)
+	lunObject.LunPath = lunPath
+	files, err := ioutil.ReadDir(lunPath)
+	if err != nil {
+		return lunObject, fmt.Errorf("getLunLinkTarget: ReadDir path %s error %v", lunPath, err)
+	}
+	for _, file := range files {
+		fileInfo, _ := os.Lstat(filepath.Join(lunPath, file.Name()))
+		if fileInfo.Mode()&os.ModeSymlink != 0 {
+			target, err := os.Readlink(filepath.Join(lunPath, fileInfo.Name()))
+			if err != nil {
+				return lunObject, fmt.Errorf("getLunLinkTarget: Readlink err %v", err)
+			}
+			targetPath, objectName := filepath.Split(target)
+			_, typeWithNumber := filepath.Split(filepath.Clean(targetPath))
+
+			splitTypeWithNumber := strings.Split(typeWithNumber, "_")
```

Comment:
I think this could be a little nicer using `strings.LastIndex`, and then using the index to split the string.
```golang
underscore := strings.LastIndex(typeWithNumber, "_")
lunObject.Backstore = typeWithNumber[:underscore]
lunObject.ObjectName = typeWithNumber[underscore+1:]
```
Might also need to check for a filename without an underscore.

Reply:
There are 2 kind of of backstore name format, 
backstore_number or back_store_number 
I guess we can error out when no underscore is being found.